storage/innobase/fil/fil0crypt.cc

```diff
@@ -967,9 +967,11 @@ static fil_space_crypt_t *fil_space_set_crypt_data(
 @param[in]  ptr  Log entry start
 @param[in]  end_ptr  Log entry end
 @param[in]  len  Log entry length
+@param[in]  recv_needed_recovery  missing keys will report error
 @return position on log buffer */
 byte *fil_parse_write_crypt_data_v3(space_id_t space_id, byte *ptr,
-                                    const byte *end_ptr, ulint len) {
+                                    const byte *end_ptr, ulint len,
+                                    bool recv_needed_recovery) {
   ptr += 4;  // skip offset and len
```

Comment:
but upstream also calls fil_tablespace_redo_encryption() always. How does it handle this ? not complain about old keys?

Reply:
Discussion from Slack:
Robert Golebiowski  1 hour ago
the question is why parse do not always complain about missing keys ?

Satya Bodapati  1 hour ago
yes, how does it choose not to complain or error out?

Robert Golebiowski  1 hour ago
I think it does error out - it just wasn't tested in upstream

Robert Golebiowski  1 hour ago
when "encryption redo" is in redo, but the first page was already flushed

Robert Golebiowski  1 hour ago
@satya - do you want me to test it ?

Satya Bodapati  1 hour ago
how do we make a key invalid? remove after rotation?

Satya Bodapati  1 hour ago
if (!Encryption::decode_encryption_info(key, iv, ptr, true)) {
      recv_sys->found_corrupt_log = true;
      ib::warn(ER_IB_MSG_364)
          << "Encryption information"
          << " in the redo log of space " << space_id << " is invalid";
      return (nullptr);
    }

Satya Bodapati  1 hour ago
in fil_tablespace_redo_encryption()

Satya Bodapati  1 hour ago
so it looks it can fail... :thinking_face:

Satya Bodapati  1 hour ago
yes, then we can create upstream bug

Satya Bodapati  1 hour ago
breaking for lunch.. bbl

Satya Bodapati  1 hour ago
back

Robert Golebiowski  43 minutes ago
kk

Robert Golebiowski  20 minutes ago
yeah, it errors out during redo processing

I have create bug report in upstream : 100491.